code action for deleting unused definitions

This is probably a good first issue, unless working out exactly to delete proves hard.

If a symbol isn't used and isn't exported, GHC emits this warning.
```
C:\msys64\home\ganesh\temp\x\Foo.hs:3:1: warning:
    Defined but not used: ‘foo’
```
It'd be great to have a code action to delete it.

Complicating factors:
 - the text range from the warning only points to the name itself, not the entire definition, so we need to find the full definition somehow. I guess GHC's parse tree should be available?
 - you get the same warning for local definitions and unused parameters. Local definitions can also just be deleted but it may be harder to work out exactly what to delete. Unused parameters should be either turned into or prefixed with '_'.

--- a/src/Development/IDE/GHC/Compat.hs
+++ b/src/Development/IDE/GHC/Compat.hs
@@ -32,6 +32,8 @@ module Development.IDE.GHC.Compat(
     pattern InstD,
     pattern TyClD,
     pattern ValD,
+    pattern SigD,
+    pattern TypeSig,
     pattern ClassOpSig,
     pattern IEThingAll,
     pattern IEThingWith,
@@ -52,7 +54,7 @@ import Packages
 
 import qualified GHC
 import GHC hiding (
-      ClassOpSig, DerivD, ForD, IEThingAll, IEThingWith, InstD, TyClD, ValD, ModLocation
+      ClassOpSig, DerivD, ForD, IEThingAll, IEThingWith, InstD, TyClD, ValD, SigD, TypeSig, ModLocation
 #if MIN_GHC_API_VERSION(8,6,0)
     , getConArgs
 #endif
@@ -156,6 +158,22 @@ pattern TyClD x <-
     GHC.TyClD _ x
 #else
     GHC.TyClD x
+#endif
+
+pattern SigD :: Sig p -> HsDecl p
+pattern SigD x <-
+#if MIN_GHC_API_VERSION(8,6,0)
+    GHC.SigD _ x
+#else
+    GHC.SigD x
+#endif
+
+pattern TypeSig :: [Located (IdP p)] -> LHsSigWcType p -> Sig p
+pattern TypeSig x y <-
+#if MIN_GHC_API_VERSION(8,6,0)
+    GHC.TypeSig _ x y
+#else
+    GHC.TypeSig x y
 #endif
 
 pattern ClassOpSig :: Bool -> [Located (IdP pass)] -> LHsSigType pass -> Sig pass

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -191,6 +191,7 @@ suggestDeleteTopBinding ParsedModule{pm_parsed_source = L _ HsModule{hsmodDecls}
                         . sortLocated
                         $ hsmodDecls
           sameName = filter (matchesBindingName (T.unpack name) . snd) allTopLevel
+    , not (null sameName)
             = [("Delete ‘" <> name <> "’", flip TextEdit "" . toNextBinding allTopLevel . fst <$> sameName )]
     | otherwise = []
     where

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -48,6 +48,7 @@ import Data.List.Extra
 import qualified Data.Text as T
 import Data.Tuple.Extra ((&&&))
 import HscTypes
+import SrcLoc
 import Parser
 import Text.Regex.TDFA ((=~), (=~~))
 import Text.Regex.TDFA.Text()
@@ -185,9 +186,11 @@ suggestDeleteTopBinding :: ParsedModule -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestDeleteTopBinding ParsedModule{pm_parsed_source = L _ HsModule{hsmodDecls}} Diagnostic{_range=_range,..}
 -- Foo.hs:4:1: warning: [-Wunused-top-binds] Defined but not used: ‘f’
     | Just [name] <- matchRegex _message ".*Defined but not used: ‘([^ ]+)’"
-    , let
-        allTopLevel = filter (isTopLevel . fst) $ map (\(L l b) -> (srcSpanToRange l, b)) hsmodDecls
-        sameName = filter (matchesBindingName (T.unpack name) . snd) allTopLevel
+    , let allTopLevel = filter (isTopLevel . fst)
+                        . map (\(L l b) -> (srcSpanToRange l, b))
+                        . sortLocated
+                        $ hsmodDecls
+          sameName = filter (matchesBindingName (T.unpack name) . snd) allTopLevel
             = [("Delete ‘" <> name <> "’", flip TextEdit "" . toNextBinding allTopLevel . fst <$> sameName )]
     | otherwise = []
     where

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -185,18 +185,24 @@ suggestDeleteTopBinding :: ParsedModule -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestDeleteTopBinding ParsedModule{pm_parsed_source = L _ HsModule{hsmodDecls}} Diagnostic{_range=_range,..}
 -- Foo.hs:4:1: warning: [-Wunused-top-binds] Defined but not used: ‘f’
     | Just [name] <- matchRegex _message ".*Defined but not used: ‘([^ ]+)’"
-    , (_character . _start) _range == 0
-    , let sameName = filter (\(L l b) -> isTopLevel l && matchesBindingName b (T.unpack name)) hsmodDecls
-            = [("Delete ‘" <> name <> "’", [TextEdit (nextLine . srcSpanToRange $ l) "" | (L l _) <- sameName])]
+    , let
+        allTopLevel = filter (isTopLevel . fst) $ map (\(L l b) -> (srcSpanToRange l, b)) hsmodDecls
+        sameName = filter (matchesBindingName (T.unpack name) . snd) allTopLevel
+            = [("Delete ‘" <> name <> "’", flip TextEdit "" . toNextBinding allTopLevel . fst <$> sameName )]
     | otherwise = []
     where
-      isTopLevel l = (_character . _start) (srcSpanToRange l) == 0
+      isTopLevel l = (_character . _start) l == 0
 
-      nextLine r = r {_end = (_end r) {_line = (_line . _end $ r) + 1, _character = 0}}
+      forwardLines lines r = r {_end = (_end r) {_line = (_line . _end $ r) + lines, _character = 0}}
 
-      matchesBindingName :: HsDecl GhcPs -> String -> Bool
-      matchesBindingName (ValD FunBind {fun_id=L _ x}) b = showSDocUnsafe (ppr x) == b
-      matchesBindingName (SigD (TypeSig (L _ x:_) _)) b = showSDocUnsafe (ppr x) == b
+      toNextBinding bindings r@Range { _end = Position {_line = l} }
+        | Just (Range { _start = Position {_line = l'}}, _) <- find ((> l) . _line . _start . fst) bindings
+        = forwardLines (l' - l) r
+      toNextBinding _ r  = r
+
+      matchesBindingName :: String -> HsDecl GhcPs -> Bool
+      matchesBindingName b (ValD FunBind {fun_id=L _ x}) = showSDocUnsafe (ppr x) == b
+      matchesBindingName b (SigD (TypeSig (L _ x:_) _)) = showSDocUnsafe (ppr x) == b
       matchesBindingName _ _ = False
 
 suggestReplaceIdentifier :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -158,8 +158,8 @@ suggestAction dflags packageExports ideOptions parsedModule text diag = concat
     [  suggestNewDefinition ideOptions pm text diag
     ++ suggestRemoveRedundantImport pm text diag
     ++ suggestNewImport packageExports pm diag
+    ++ suggestDeleteTopBinding pm diag
     | Just pm <- [parsedModule]]
-
 
 suggestRemoveRedundantImport :: ParsedModule -> Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestRemoveRedundantImport ParsedModule{pm_parsed_source = L _  HsModule{hsmodImports}} contents Diagnostic{_range=_range,..}
@@ -179,6 +179,24 @@ suggestRemoveRedundantImport ParsedModule{pm_parsed_source = L _  HsModule{hsmod
     | _message =~ ("The( qualified)? import of [^ ]* is redundant" :: String)
         = [("Remove import", [TextEdit (extendToWholeLineIfPossible contents _range) ""])]
     | otherwise = []
+
+suggestDeleteTopBinding :: ParsedModule -> Diagnostic -> [(T.Text, [TextEdit])]
+suggestDeleteTopBinding ParsedModule{pm_parsed_source = L _ HsModule{hsmodDecls}} Diagnostic{_range=_range,..}
+-- Foo.hs:4:1: warning: [-Wunused-top-binds] Defined but not used: ‘f’
+    | Just [name] <- matchRegex _message ".*Defined but not used: ‘([^ ]+)’"
+    , (_character . _start) _range == 0
+    , let sameName = filter (\(L l b) -> isTopLevel l && matchesBindingName b (T.unpack name)) hsmodDecls
+            = [("Delete ‘" <> name <> "’", [TextEdit (nextLine . srcSpanToRange $ l) "" | (L l _) <- sameName])]
+    | otherwise = []
+    where
+      isTopLevel l = (_character . _start) (srcSpanToRange l) == 0
+
+      nextLine r = r {_end = (_end r) {_line = (_line . _end $ r) + 1, _character = 0}}
+
+      matchesBindingName :: HsDecl GhcPs -> String -> Bool
+      matchesBindingName (ValD (FunBind {fun_id=L _ x})) b = showSDocUnsafe (ppr x) == b
+      matchesBindingName (SigD _ (TypeSig _ (L _ x:_) _)) b = showSDocUnsafe (ppr x) == b
+      matchesBindingName _ _ = False
 
 suggestReplaceIdentifier :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestReplaceIdentifier contents Diagnostic{_range=_range,..}

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -161,6 +161,7 @@ suggestAction dflags packageExports ideOptions parsedModule text diag = concat
     ++ suggestDeleteTopBinding pm diag
     | Just pm <- [parsedModule]]
 
+
 suggestRemoveRedundantImport :: ParsedModule -> Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]
 suggestRemoveRedundantImport ParsedModule{pm_parsed_source = L _  HsModule{hsmodImports}} contents Diagnostic{_range=_range,..}
 --     The qualified import of ‘many’ from module ‘Control.Applicative’ is redundant
@@ -195,7 +196,7 @@ suggestDeleteTopBinding ParsedModule{pm_parsed_source = L _ HsModule{hsmodDecls}
 
       matchesBindingName :: HsDecl GhcPs -> String -> Bool
       matchesBindingName (ValD FunBind {fun_id=L _ x}) b = showSDocUnsafe (ppr x) == b
-      matchesBindingName (SigD _ (TypeSig _ (L _ x:_) _)) b = showSDocUnsafe (ppr x) == b
+      matchesBindingName (SigD (TypeSig (L _ x:_) _)) b = showSDocUnsafe (ppr x) == b
       matchesBindingName _ _ = False
 
 suggestReplaceIdentifier :: Maybe T.Text -> Diagnostic -> [(T.Text, [TextEdit])]

--- a/src/Development/IDE/Plugin/CodeAction.hs
+++ b/src/Development/IDE/Plugin/CodeAction.hs
@@ -194,7 +194,7 @@ suggestDeleteTopBinding ParsedModule{pm_parsed_source = L _ HsModule{hsmodDecls}
       nextLine r = r {_end = (_end r) {_line = (_line . _end $ r) + 1, _character = 0}}
 
       matchesBindingName :: HsDecl GhcPs -> String -> Bool
-      matchesBindingName (ValD (FunBind {fun_id=L _ x})) b = showSDocUnsafe (ppr x) == b
+      matchesBindingName (ValD FunBind {fun_id=L _ x}) b = showSDocUnsafe (ppr x) == b
       matchesBindingName (SigD _ (TypeSig _ (L _ x:_) _)) b = showSDocUnsafe (ppr x) == b
       matchesBindingName _ _ = False
 

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -1148,35 +1148,63 @@ insertNewDefinitionTests = testGroup "insert new definition actions"
 
 deleteUnusedDefinitionTests :: TestTree
 deleteUnusedDefinitionTests = testGroup "delete unused definition action"
-  [ testSession "delete unused top level binding" $ do
-      let source = T.unlines [ "{-# OPTIONS_GHC -Wunused-top-binds #-}"
-                             , "module A (some) where"
-                             , ""
-                             , "f :: Int -> Int"
-                             , "f 1 = let a = 1"
-                             , "      in a"
-                             , "f 2 = 2"
-                             , ""
-                             , "some = ()"
-                             ]
-      docId <- createDoc "A.hs" "haskell" source
-      expectDiagnostics [ ("A.hs", [(DsWarning, (4, 0), "not used")]) ]
-
-      Just (CACodeAction action@CodeAction { _title = actionTitle })
-                  <- find (\(CACodeAction CodeAction{_title=x}) -> "Delete" `T.isPrefixOf` x )
-                     <$> getCodeActions docId (R 0 0 0 0)
-
-      liftIO $ actionTitle @?= "Delete ‘f’"
-      executeCodeAction action
-      contentAfterAction <- documentContents docId
-      liftIO $ contentAfterAction @?= T.unlines [
+  [ testSession "delete unused top level binding" $
+    testFor
+    (T.unlines [ "{-# OPTIONS_GHC -Wunused-top-binds #-}"
+               , "module A (some) where"
+               , ""
+               , "f :: Int -> Int"
+               , "f 1 = let a = 1"
+               , "      in a"
+               , "f 2 = 2"
+               , ""
+               , "some = ()"
+               ])
+    (4, 0)
+    "Delete ‘f’"
+    (T.unlines [
         "{-# OPTIONS_GHC -Wunused-top-binds #-}"
         , "module A (some) where"
         , ""
+        , "some = ()"
+        ])
+
+  , testSession "delete unused top level binding defined in infix form" $
+    testFor
+    (T.unlines [ "{-# OPTIONS_GHC -Wunused-top-binds #-}"
+               , "module A (some) where"
+               , ""
+               , "myPlus :: Int -> Int -> Int"
+               , "a `myPlus` b = a + b"
+               , ""
+               , "some = ()"
+               ])
+    (4, 2)
+    "Delete ‘myPlus’"
+    (T.unlines [
+        "{-# OPTIONS_GHC -Wunused-top-binds #-}"
+        , "module A (some) where"
         , ""
         , "some = ()"
-        ]
+      ])
   ]
+  where
+    testFor source pos expectedTitle expectedResult = do
+      docId <- createDoc "A.hs" "haskell" source
+      expectDiagnostics [ ("A.hs", [(DsWarning, pos, "not used")]) ]
+
+      (action, title) <- extractCodeAction docId "Delete"
+
+      liftIO $ title @?= expectedTitle
+      executeCodeAction action
+      contentAfterAction <- documentContents docId
+      liftIO $ contentAfterAction @?= expectedResult
+
+    extractCodeAction docId actionPrefix = do
+      Just (CACodeAction action@CodeAction { _title = actionTitle })
+                  <- find (\(CACodeAction CodeAction{_title=x}) -> actionPrefix `T.isPrefixOf` x)
+                     <$> getCodeActions docId (R 0 0 0 0)
+      return (action, actionTitle)
 
 
 fixConstructorImportTests :: TestTree

--- a/test/exe/Main.hs
+++ b/test/exe/Main.hs
@@ -476,6 +476,7 @@ codeActionTests = testGroup "code actions"
   , fillTypedHoleTests
   , addSigActionTests
   , insertNewDefinitionTests
+  , deleteUnusedDefinitionTests
   ]
 
 codeLensesTests :: TestTree
@@ -1143,6 +1144,40 @@ insertNewDefinitionTests = testGroup "insert new definition actions"
         ]
         ++ txtB')
   ]
+
+
+deleteUnusedDefinitionTests :: TestTree
+deleteUnusedDefinitionTests = testGroup "delete unused definition action"
+  [ testSession "delete unused top level binding" $ do
+      let source = T.unlines [ "{-# OPTIONS_GHC -Wunused-top-binds #-}"
+                             , "module A (some) where"
+                             , ""
+                             , "f :: Int -> Int"
+                             , "f 1 = let a = 1"
+                             , "      in a"
+                             , "f 2 = 2"
+                             , ""
+                             , "some = ()"
+                             ]
+      docId <- createDoc "A.hs" "haskell" source
+      expectDiagnostics [ ("A.hs", [(DsWarning, (4, 0), "not used")]) ]
+
+      Just (CACodeAction action@CodeAction { _title = actionTitle })
+                  <- find (\(CACodeAction CodeAction{_title=x}) -> "Delete" `T.isPrefixOf` x )
+                     <$> getCodeActions docId (R 0 0 0 0)
+
+      liftIO $ actionTitle @?= "Delete ‘f’"
+      executeCodeAction action
+      contentAfterAction <- documentContents docId
+      liftIO $ contentAfterAction @?= T.unlines [
+        "{-# OPTIONS_GHC -Wunused-top-binds #-}"
+        , "module A (some) where"
+        , ""
+        , ""
+        , "some = ()"
+        ]
+  ]
+
 
 fixConstructorImportTests :: TestTree
 fixConstructorImportTests = testGroup "fix import actions"

